kratos/linear_solvers/amgcl_solver.h

```diff
@@ -382,18 +382,72 @@ class AMGCLSolver : public LinearSolver< TSparseSpaceType,
             mAMGCLParameters.put("precond.coarse_enough",mCoarseEnough/mBlockSize);
 
         Matrix B;
-        if(mUseAMGPreconditioning && mProvideCoordinates) {
-            B = ZeroMatrix(  TSparseSpaceType::Size1(rA), mBlockSize*4  );
-            for(IndexType i=0; i<TSparseSpaceType::Size1(rA); i+=mBlockSize) {
-                for( IndexType j=0; j<static_cast<IndexType>(mBlockSize); j++) {
-                    B(i+j,  j) = 1.0;
-
-                    IndexType inode = i/mBlockSize;
+        if(mUseAMGPreconditioning && mProvideCoordinates && (mBlockSize == 2 || mBlockSize == 3)) {
+            double sn = sqrt(TSparseSpaceType::Size1(rA));
+
+            if (mBlockSize == 2) {
+                B = ZeroMatrix(  TSparseSpaceType::Size1(rA), 3);
+                for(IndexType i = 0; i<TSparseSpaceType::Size1(rA); ++i) {
+                    IndexType inode = i / mBlockSize;
+                    IndexType coord = i % mBlockSize;
+
+                    // Translation
+                    B(i,coord) = sn;
+
+                    // Rotation
+                    switch(coord) {
+                        case 0:
+                            B(i,2) = -mCoordinates[inode][1];
+                            break;
+                        case 1:
+                            B(i,2) = mCoordinates[inode][0];
+                            break;
+                    }
+                }
+            } else if (mBlockSize == 3) {
+                B = ZeroMatrix(  TSparseSpaceType::Size1(rA), 6);
+                for(IndexType i = 0; i<TSparseSpaceType::Size1(rA); ++i) {
+                    IndexType inode = i / mBlockSize;
+                    IndexType coord = i % mBlockSize;
+
+                    // Translation
+                    B(i,coord) = sn;
+
+                    // Rotation
+                    switch(coord) {
+                        case 0:
+                            B(i,3) = mCoordinates[inode][1];
+                            B(i,5) = mCoordinates[inode][2];
+                            break;
+                        case 1:
+                            B(i,3) = -mCoordinates[inode][0];
+                            B(i,4) = -mCoordinates[inode][2];
+                            break;
+                        case 2:
+                            B(i,4) = mCoordinates[inode][1];
+                            B(i,5) = -mCoordinates[inode][0];
+                            break;
+                    }
+                }
+            }
 
-                    B(i+j, mBlockSize +j*3 + 0) = mCoordinates[inode][0];
-                    B(i+j, mBlockSize +j*3 + 1) = mCoordinates[inode][1];
-                    B(i+j, mBlockSize +j*3 + 2) = mCoordinates[inode][2];
+            // Orthonormalization
+            std::vector<double> dot(B.size2());
+            for(IndexType i = mBlockSize; i < B.size2(); ++i) {
+                std::fill(dot.begin(), dot.end(), 0.0);
+                for(IndexType j = 0; j < TSparseSpaceType::Size1(rA); ++j) {
```

Comment:
The loops here could benefit from OpenMP pragmas, not sure what is the Kratos policy regarding this.